### DISCUSSION TITLE
Link 2 documents from 2 collections of the same data base from MongoDB in Next.js

### DISCUSSION THREAD

**Discussion by mihaimazareanu** [neuefische/questions#386] [Web] [answered]
Hi,

I have 1 data base in MongoDB which has 2 collections, `parents` and `children`. The 2 collections DON'T have any documents that are the same. All documents from both collections have an ID, first name and last name, but they don't have the same values.

All documents in the collection `parents` have a property `children` and this is where I want to insert the documents from the `children` collection.

Basically I want that when a parent creates an account for their children, that account will be added in the `children` collection (already successfully implemented) and basically the parent account should somehow be linked with this newly created child account, because I need the parents to manage the child accounts for further functionality of my app.

Any idea how I can implement this in my Next.js app?

From the research I have done so far, I assume I have to use the aggregate() function, but I have no idea which operator I need, as $lookup needs a matching field in both collections, if I understood everything correctly and I don't have any matching fields...

Help would be highly appreciated here, as I can't move on with the next user stories until I have implemented this... 🙏

**Comment by actyvystom** [comment]
Hi Mihai,
you are on the right track 😉 

1. You need to alter your child mongoose model and add an id for the parent as an ObjectId-type (e.g. `parentId`)

```js
...
parentId: {
    type: mongoose.Types.ObjectId,
  },
...
``` 
2. You need to alter your POST request when adding children to your parents account and add this `parentId` to your data body  
3. Then you can change your api GET route for your users and implement an aggregation query with a `$lookup` stage, e.g.:

```js
...
<yourparentcollection>.aggregate([
    
  {
    $lookup: {
      from: 'children',
      localField: '_id',
      foreignField: 'parentId',
      as: 'children',
    },
  }
]).exec()
...

```
 Hope that helps! 🤞🏻

**Reply by mihaimazareanu** [reply]
Thanks for your help, Thomas, but I can't seem to get the code right... 😭 

I have written this code in my `api/parents.js` file:

    import connectDB from "./_db/connect-db";
    import {Parent} from "./_db/models/Parent";

    const mongodb = require("mongodb");

    const url =
    "mongodb+srv://mihaimazareanu:<password>@kidsfi.jnbdwx0.mongodb.net/test";

    async function handler(req, res) {
    const client = await mongodb.MongoClient.connect(url);
    const db = client.db();
    const parents = db.collection("parents");
    switch (req.method) {
    case "GET":
      try {
        const filter = {};
        if (req.query.firstName) {
          filter.firstName = req.query.firstName;
          // filter.lastName = req.query.lastName;
          // filter.password = req.query.password;
        }
        const users = await Parent.find(filter);

        const children = await parents
          .aggregate([
            {
              $lookup: {
                from: "children",
                localField: "_id",
                foreignField: "parentID",
                as: "children",
              },
            },
            {
              $project: {
                children: 1,
              },
            },
          ])
          .toArray();
        res.status(200).json(users, children);
      } catch (error) {
        // You can inspect the error and return more meaningful error messages...
        res.status(500).json({error: "something went wrong"});
      }
      break;
    case "POST":
      try {
        const body = req.body;
        const newParent = new Parent({
          firstName: body.firstName,
          lastName: body.lastName,
          password: body.password,
          isParent: body.isParent,
        });
        await newParent.save();
        res.status(200).json(newParent);
      } catch (error) {
        res.status(500).json({error: error.message});
      }
      break;
    case "DELETE":
    case "PUT":
    case "PATCH":
    default:
      res.status(405).json({error: "method not allowed"});
    }
    }
    
    export default connectDB(handler);
    
However, when I look into MongoDB, the document in the collection `parents` doesn't get the new field `children`, as stated in the pipeline and also I console.log()-ed the parent user in the app and the object doesn't have the property `children` either.

So obviously I am doing it wrong, but after almost 4 hours at it, I am running out of ideas...

I would appreciate some (more) help here... 🙈 

PS: I have pushed the changes in the PR, so you can see the whole code here:

https://github.com/mihaimazareanu/capstone-project-kidsfi/pull/22

**Comment by actyvystom** [comment, marked as answer]
Hi Mihai,

there are some issues with your code and obviously due to a misunderstanding of the aggregation functionality. The collections of `parents` and `children` are not modified by an aggregation, it's a way of keeping them separate but linking / joining them together when needed. 
That said, the result of an aggregation command is **just virtual** and only present in your code, you won't see it in your mongo DB! 

So, an aggregation is made up of phases/stages you can individually plug together. So, to get the parent user who is logging in, you need to add a `$match`-stage to your aggregation pipeline, **before** your `$lookup`-stage (which will replace your `.find()`-query on `parents`!) e.g.:

```js
const parents = await Parent.aggregate([
          {
            $match: {
              firstName: req.query.firstName,
            },
          },
          {
            $lookup: {
              from: "children",
              localField: "_id",
              foreignField: "parentID",
              as: "children",
            },
          },
        ]).exec();
```
You don't need a `$projection`-stage as you already define in your `$lookup` where the children documents should go.
**Very important is the `.exec()`-statement at the end of the aggregate-pipeline, it's needed to `execute` the db query** - so you have to remove your `.toArray()`-method at the end of the statement and replace it with `.exec()`

Once executed your data will be available in `parents` which you can output as known:

```js
res.status(200).json(parents);
``` 
Hope that helps!

**Reply by mihaimazareanu** [reply]
I think I understand better now how the aggregate function works. The materials that I found online weren't very helpful, unfortunately, so there was quite a lot of guessing from my part... But my aggregate function works fine now, I only need to do some fine tuning to my code, as it acts weird after creating a child account, but I don't believe it has anything to do with the aggregate function, as the children are being console.logged. But that's something for tomorrow. 😀

Thank you for taking your time to help me, especially on a Saturday. 🙏
